# Answering Business Questions using SQL

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [3]:
%%sql
--finding the total number of tracks that were sold within the United States

SELECT COUNT(quantity) total_tracks_sold_usa FROM invoice_line il
INNER JOIN invoice i ON i.invoice_id = il.invoice_id
INNER JOIN customer c ON c.customer_id = i.customer_id
WHERE c.country = "USA"

 * sqlite:///chinook.db
Done.


[(1051,)]

In [4]:
%%sql
--now, we need to dig deeper and break down the total number of tracks sold in the usa by genre and market share
--Write a query that returns each genre, with the number of tracks sold in the USA: in absolute numbers; percentages

WITH 
    usa AS 
    (
    SELECT il.*, i.*, c.* FROM invoice_line il
    INNER JOIN invoice i ON i.invoice_id = il.invoice_id
    INNER JOIN customer c ON c.customer_id = i.customer_id
    WHERE c.country = "USA"
    )
    
SELECT g.name genre, SUM(quantity) tracks_sold_usa, CAST(count(quantity) as float)/ 
    ( 
    SELECT COUNT(*) FROM usa 
    ) percentage_sold       
FROM usa
INNER JOIN track t ON t.track_id = usa.track_id
INNER JOIN genre g ON g.genre_id = t.genre_id
GROUP BY 1
ORDER BY 2 DESC

 * sqlite:///chinook.db
Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Latin', 22, 0.02093244529019981),
 ('Pop', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425),
 ('Easy Listening', 13, 0.012369172216936251),
 ('Reggae', 6, 0.005708848715509039),
 ('Electronica/Dance', 5, 0.004757373929590866),
 ('Classical', 4, 0.003805899143672693),
 ('Heavy Metal', 3, 0.0028544243577545195),
 ('Soundtrack', 2, 0.0019029495718363464),
 ('TV Shows', 1, 0.0009514747859181732)]

#### If a music shop in the USA wanted to restock on inventory, they would be wise to buy a relatively larger share of Rock, Alternative & Punk, and Metal albums in comparison to other genres.

---

In [5]:
%%sql
--Write a query that finds the total $ amount of sales assigned to each sales support agent within the company
--Compare the results of the sales support agents' performances

WITH support_agent_info AS
    (
    SELECT 
        e.employee_id,
        e.first_name || " " || e.last_name employee_name,
        e.hire_date,
        ROUND(SUM(i.total), 2) total_sales
    FROM employee e
    INNER JOIN customer c ON c.support_rep_id = e.employee_id
    INNER JOIN invoice i ON i.customer_id = c.customer_id
    GROUP BY 1, 2
    )
    
SELECT * FROM support_agent_info sai
GROUP BY 1

 * sqlite:///chinook.db
Done.


[(3, 'Jane Peacock', '2017-04-01 00:00:00', 1731.51),
 (4, 'Margaret Park', '2017-05-03 00:00:00', 1584.0),
 (5, 'Steve Johnson', '2017-10-17 00:00:00', 1393.92)]

#### While there is a 20% difference in sales between Jane (the top employee) and Steve (the bottom employee), the difference roughly corresponds with the differences in their hiring dates.

---

In [6]:
%%sql
--Write a query that collates data on purchases from different countries
--Where a country has only one customer, they will be sorted into an "other" group
--The results are sorted by the total sales in descending order, with the "other" group at the very bottom

WITH 
    country_list AS
    (
     SELECT
        CASE
            WHEN
                (
                SELECT COUNT(*) FROM customer
                WHERE country = c.country
                ) = 1
                THEN "Other"
                ELSE c.country
            END as country,
        c.customer_id,
        il.*
     FROM invoice_line il
     INNER JOIN invoice i ON i.invoice_id = il.invoice_id
     INNER JOIN customer c ON c.customer_id = i.customer_id
     ORDER BY country ASC
    )
           
SELECT 
    country,
    total_customers,
    total_sales,
    average_order,
    customer_lifetime_value

FROM 
    (
     SELECT
        country,
        count(distinct customer_id) total_customers,
        SUM(unit_price) total_sales,
        SUM(unit_price) / count(distinct customer_id) customer_lifetime_value,
        SUM(unit_price) / count(distinct invoice_id) average_order,
        CASE
            WHEN country = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM country_list
    GROUP BY country
    ORDER BY sort ASC, total_sales DESC
    );

 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.490000000008, 7.942671755725252, 80.03769230769292),
 ('Canada', 8, 535.5900000000034, 7.047236842105309, 66.94875000000043),
 ('Brazil', 5, 427.68000000000245, 7.011147540983647, 85.53600000000048),
 ('France', 5, 389.0700000000021, 7.781400000000042, 77.81400000000042),
 ('Germany', 4, 334.6200000000016, 8.161463414634186, 83.6550000000004),
 ('Czech Republic', 2, 273.24000000000103, 9.108000000000034, 136.62000000000052),
 ('United Kingdom', 3, 245.52000000000078, 8.768571428571457, 81.84000000000026),
 ('Portugal', 2, 185.13000000000022, 6.3837931034482835, 92.56500000000011),
 ('India', 2, 183.1500000000002, 8.72142857142858, 91.5750000000001),
 ('Other', 15, 1094.9400000000085, 7.448571428571486, 72.99600000000056)]

#### If we were debating whether to spend money on an advertising campaign, we would have to be cautious about investing too much money. There quantity of data from each of these countries is relatively low; with a small sample size, we need to be wary of the reliability of the data. The best approach would be to run small campaigns in these countries, collecting and analyzing the new customers to make sure that these trends hold with new customers.

---

In [9]:
%%sql
--query to see the album_id and number of purchases
--since there are no 0s, that means that every album in database was bought at least once

WITH 
    num_of_album_tracks_sold AS 
        (
        SELECT 
            album_id,
            COUNT(track_id) num_tracks_sold
        FROM track
        GROUP BY 1
        )
    
SELECT album_id, num_tracks_sold FROM num_of_album_tracks_sold

 * sqlite:///chinook.db
Done.


[(1, 10),
 (2, 1),
 (3, 3),
 (4, 8),
 (5, 15),
 (6, 13),
 (7, 12),
 (8, 14),
 (9, 8),
 (10, 14),
 (11, 12),
 (12, 12),
 (13, 8),
 (14, 13),
 (15, 5),
 (16, 7),
 (17, 10),
 (18, 17),
 (19, 11),
 (20, 11),
 (21, 18),
 (22, 3),
 (23, 34),
 (24, 23),
 (25, 13),
 (26, 17),
 (27, 14),
 (28, 10),
 (29, 14),
 (30, 14),
 (31, 9),
 (32, 14),
 (33, 17),
 (34, 17),
 (35, 11),
 (36, 17),
 (37, 20),
 (38, 12),
 (39, 21),
 (40, 12),
 (41, 14),
 (42, 14),
 (43, 7),
 (44, 6),
 (45, 14),
 (46, 13),
 (47, 14),
 (48, 13),
 (49, 10),
 (50, 4),
 (51, 22),
 (52, 15),
 (53, 14),
 (54, 20),
 (55, 20),
 (56, 15),
 (57, 15),
 (58, 9),
 (59, 7),
 (60, 7),
 (61, 11),
 (62, 7),
 (63, 12),
 (64, 9),
 (65, 9),
 (66, 10),
 (67, 16),
 (68, 9),
 (69, 13),
 (70, 13),
 (71, 14),
 (72, 18),
 (73, 30),
 (74, 12),
 (75, 14),
 (76, 15),
 (77, 11),
 (78, 14),
 (79, 10),
 (80, 10),
 (81, 11),
 (82, 13),
 (83, 24),
 (84, 16),
 (85, 14),
 (86, 15),
 (87, 3),
 (88, 12),
 (89, 13),
 (90, 12),
 (91, 16),
 (92, 14),
 (93, 13),
 (94, 11),
 (95, 12),
 (96, 11),
 (97, 10),
 (98, 11),
 (99, 12),
 (100, 9),
 (101, 10),
 (102, 18),
 (103, 10),
 (104, 10),
 (105, 10),
 (106, 9),
 (107, 8),
 (108, 10),
 (109, 9),
 (110, 8),
 (111, 8),
 (112, 8),
 (113, 11),
 (114, 8),
 (115, 20),
 (116, 10),
 (117, 11),
 (118, 11),
 (119, 13),
 (120, 17),
 (121, 10),
 (122, 14),
 (123, 12),
 (124, 14),
 (125, 16),
 (126, 15),
 (127, 10),
 (128, 8),
 (129, 8),
 (130, 7),
 (131, 8),
 (132, 9),
 (133, 9),
 (134, 10),
 (135, 9),
 (136, 7),
 (137, 5),
 (138, 4),
 (139, 15),
 (140, 16),
 (141, 57),
 (142, 14),
 (143, 14),
 (144, 10),
 (145, 18),
 (146, 18),
 (147, 10),
 (148, 12),
 (149, 16),
 (150, 10),
 (151, 14),
 (152, 8),
 (153, 13),
 (154, 8),
 (155, 11),
 (156, 9),
 (157, 14),
 (158, 13),
 (159, 13),
 (160, 15),
 (161, 12),
 (162, 17),
 (163, 17),
 (164, 12),
 (165, 15),
 (166, 14),
 (167, 21),
 (168, 12),
 (169, 16),
 (170, 1),
 (171, 2),
 (172, 1),
 (173, 2),
 (174, 14),
 (175, 12),
 (176, 14),
 (177, 10),
 (178, 16),
 (179, 13),
 (180, 15),
 (181, 11),
 (182, 12),
 (183, 9),
 (184, 16),
 (185, 17),
 (186, 11),
 (187, 11),
 (188, 11),
 (189, 14),
 (190, 16),
 (191, 10),
 (192, 14),
 (193, 17),
 (194, 16),
 (195, 15),
 (196, 14),
 (197, 8),
 (198, 6),
 (199, 12),
 (200, 11),
 (201, 16),
 (202, 18),
 (203, 17),
 (204, 9),
 (205, 10),
 (206, 12),
 (207, 11),
 (208, 7),
 (209, 10),
 (210, 9),
 (211, 18),
 (212, 12),
 (213, 18),
 (214, 11),
 (215, 14),
 (216, 12),
 (217, 14),
 (218, 15),
 (219, 15),
 (220, 11),
 (221, 20),
 (222, 15),
 (223, 15),
 (224, 22),
 (225, 16),
 (226, 1),
 (227, 19),
 (228, 23),
 (229, 26),
 (230, 25),
 (231, 24),
 (232, 12),
 (233, 11),
 (234, 15),
 (235, 11),
 (236, 12),
 (237, 17),
 (238, 14),
 (239, 10),
 (240, 10),
 (241, 14),
 (242, 12),
 (243, 17),
 (244, 11),
 (245, 12),
 (246, 13),
 (247, 15),
 (248, 19),
 (249, 6),
 (250, 22),
 (251, 25),
 (252, 1),
 (253, 24),
 (254, 1),
 (255, 23),
 (256, 12),
 (257, 12),
 (258, 19),
 (259, 17),
 (260, 1),
 (261, 17),
 (262, 2),
 (263, 2),
 (264, 2),
 (265, 2),
 (266, 1),
 (267, 1),
 (268, 1),
 (269, 10),
 (270, 14),
 (271, 14),
 (272, 1),
 (273, 1),
 (274, 1),
 (275, 1),
 (276, 1),
 (277, 1),
 (278, 1),
 (279, 1),
 (280, 2),
 (281, 1),
 (282, 1),
 (283, 1),
 (284, 1),
 (285, 1),
 (286, 1),
 (287, 1),
 (288, 1),
 (289, 1),
 (290, 1),
 (291, 1),
 (292, 1),
 (293, 1),
 (294, 1),
 (295, 1),
 (296, 1),
 (297, 1),
 (298, 1),
 (299, 1),
 (300, 1),
 (301, 1),
 (302, 1),
 (303, 1),
 (304, 1),
 (305, 1),
 (306, 1),
 (307, 1),
 (308, 1),
 (309, 1),
 (310, 1),
 (311, 1),
 (312, 1),
 (313, 1),
 (314, 2),
 (315, 1),
 (316, 1),
 (317, 1),
 (318, 1),
 (319, 1),
 (320, 1),
 (321, 12),
 (322, 11),
 (323, 1),
 (324, 1),
 (325, 1),
 (326, 1),
 (327, 1),
 (328, 1),
 (329, 1),
 (330, 1),
 (331, 1),
 (332, 1),
 (333, 1),
 (334, 1),
 (335, 1),
 (336, 1),
 (337, 1),
 (338, 1),
 (339, 1),
 (340, 1),
 (341, 1),
 (342, 1),
 (343, 1),
 (344, 1),
 (345, 1),
 (346, 1),
 (347, 1)]

In [8]:
%%sql
--Write a query that categorizes each invoice as either an album purchase or not
--Within the query, calculate the number of invoices and percentage of invoices

WITH album_or_indiv AS
    (
    SELECT
        CASE
            WHEN COUNT(distinct album_id) = 1 and COUNT(il.track_id)> 3 THEN "album"
            ELSE "individual"
        END purchase_type,
        t.album_id,
        t.name,
        il.invoice_id
    FROM invoice_line il
    INNER JOIN track t ON t.track_id=il.track_id
    GROUP BY 4
    )

SELECT
    CASE
        WHEN purchase_type="album" THEN "yes"
        WHEN purchase_type="individual" THEN "no"
        END album_purchase,
    COUNT(invoice_id) total_type,
    CAST(COUNT(album_id) as float)/
            (
             SELECT COUNT(*) FROM invoice
            ) percentage
    FROM album_or_indiv ai
GROUP BY purchase_type

 * sqlite:///chinook.db
Done.


[('yes', 132, 0.21498371335504887), ('no', 482, 0.7850162866449512)]

#### Since album purchases account for nearly 21.5% of purchases, I would recommend against purchasing only select tracks from record companies;  21.5% of purchases is over one fifth of total revenue.